static/js/code-highlighter.js

```diff
@@ -169,364 +169,187 @@ var Sticky = new (class Sticky {
-      //make an array of integers and sort it for the remaining single lines
-      highlights = highlights.split(",");
-      for (var h = 0; h < highlights.length; h++) {
-        highlights[h] = parseInt(highlights[h], 10);
-        //in case some unwanted string snuck by remove it
-        if (isNaN(highlights[h])) {
-          highlights.splice(h, 1);
-          h--;
+    let line = parseInt(event.target.id.substring(1), 10);
+    if (event.shiftKey) {
+      // Range-select on shiftkey.
+      //
+      // TODO(emilio): This should maybe toggle instead of just adding to the
+      // selection?
+      if (!this.lastSelectedLine) {
+        this.addSelectedLine(line);
+      } else if (this.lastSelectedLine == line) {
+        this.removeSelectedLine(line);
+      } else if (this.lastSelectedLine < line) {
+        for (let i = this.lastSelectedLine; i < line; ++i) {
+          this.addSelectedLine(i + 1);
+        }
+      } else {
+        for (let i = this.lastSelectedLine; i > line; --i) {
+          this.addSelectedLine(i - 1);
         }
       }
-      highlights = highlights.sort(sortAsc);
-      //set the global singleLinesArry for reuse
-      singleLinesArray = highlights;
-    } else {
-      //this happens if there is no single line in a url
-      //without setting this the url gets an NaN element in it
-      highlights = null;
-    }
-
-    //a url can look like foo#12,15,20-25 or foo#12-15,18,20-25 or foo#1,2,3 etc.
-    //the lineStart should be the smallest integer in the single or highlighted ranges
-    //this ensures a proper position to which to scroll once the page loads
-    if (firstRange !== null && highlights !== null) {
-      if (highlights[0] < firstRange[0]) {
-        lineStart = highlights[0];
-      } else if (highlights[0] > firstRange[0]) {
-        lineStart = firstRange[0];
-      }
-    } else if (firstRange !== null && highlights === null) {
-      lineStart = firstRange[0];
-    } else if (firstRange === null && highlights !== null) {
-      lineStart = highlights[0];
+    } else if (event.ctrlKey || event.metaKey) {
+      // Toggle select on ctrl / meta.
+      this.toggleSelectedLine(line);
     } else {
-      lineStart = null;
+      this.selectSingleLine(line);
     }
-
-    return {
-      lineStart: lineStart,
-      highlights: singleLinesArray,
-      ranges: rangesArray,
-    };
+    this.updateHash();
   }
 
-  //first bind to all .line-number click events only
-  container.on("click", ".line-number", function (event) {
-    // hacky logic to jump if this is a stuck line number
-    if (!event.shiftKey && !event.ctrlKey) {
-      var containingLine = $(this).closest(".source-line-with-number")[0];
-      if (containingLine) {
-        if (containingLine.classList.contains("stuck")) {
-          document.getElementById("scrolling").scrollTop -=
-            containingLine.offsetTop;
-          return;
-        }
+  updateHash() {
+    let hash = this.toHash();
+    {
```

Comment:
why the extra block scope here?

Reply:
Didn't want to have the `locationHash` variable in scope unnecessarily for too long, as it may be easy to confuse it with `hash`, no big deal other than that though.

Reply:
Reasonable, thanks!